dali/c_api/c_api_test.cu

```diff
@@ -0,0 +1,170 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "dali/c_api.h"
+#include "dali/core/cuda_stream_pool.h"
+#include "dali/core/dev_buffer.h"
+#include "dali/pipeline/pipeline.h"
+
+
+namespace dali {
+
+
+namespace {
+
+__global__ void hog(float *f, size_t n) {
+  size_t i = blockIdx.x * blockDim.x + threadIdx.x;
+  if (i < n) {
+    for (int k = 0; k < 100; k++)
+      f[i] = sqrt(f[i]);
+  }
+}
+
+struct GPUHog {
+  void init() {
+    if (!mem)
+      mem = mm::alloc_raw_unique<float, mm::memory_kind::device>(size);
+    CUDA_CALL(cudaMemset(mem.get(), 1, size * sizeof(float)));
+  }
+
+  void run(cudaStream_t stream, int count = 1) {
+    for (int i = 0; i < count; i++) {
+      hog<<<div_ceil(size, 512), 512, 0, stream>>>(mem.get(), size);
+    }
+    CUDA_CALL(cudaGetLastError());
+  }
+
+  mm::uptr<float> mem;
+  size_t size = 16<<20;
+};
+
+}  // namespace
+
+enum class Method {
+  Contiguous, Samples
+};
+
+void TestCopyOutput(Method method) {
+  int batch_size = 2;
+  dali::Pipeline pipe(batch_size, 4, 0);
+  std::string es_cpu_name = "pipe_in";
+  pipe.AddExternalInput(es_cpu_name, "cpu");
+
+  std::string cont_name = "pipe_out";
+  pipe.AddOperator(OpSpec("MakeContiguous")
+                          .AddArg("device", "mixed")
+                          .AddArg("name", cont_name)
+                          .AddInput(es_cpu_name, "cpu")
+                          .AddOutput(cont_name, "gpu"), cont_name);
+  std::vector<std::pair<std::string, std::string>> outputs = {{"pipe_out", "gpu"}};
+
+  GPUHog hog;
+  hog.init();
+
+  std::vector<std::vector<int>> in_data;
+  int sample_size = 400000;
+  TensorListShape<> shape = uniform_list_shape(8, {sample_size});
+  in_data.resize(2);
+  for (int i = 0; i < 2; i++) {
+    in_data[i].resize(batch_size*sample_size);
+    for (int j = 0; j < batch_size*sample_size; j++)
+      in_data[i][j] = i + j;
+  }
+  pipe.SetOutputDescs(outputs);
+
+  std::vector<int> out_cpu(batch_size*sample_size);
+  DeviceBuffer<int> out_gpu;
+  out_gpu.resize(batch_size*sample_size);
+  CUDAStreamLease stream = CUDAStreamPool::instance().Get();
+
+  CUDA_CALL(cudaMemsetAsync(out_gpu.data(), -1, out_gpu.size() * sizeof(int), stream));
+  std::string ser = pipe.SerializeToProtobuf();
+  hog.run(stream, 50);
+
+
+  // This loop is tuned so that if the output buffer is recycled before the asynchronous copy
+  // finishes, the buffer is clobbered and an error is detected.
+  // (michalz) Verified on my desktop. The changes in c_api that came with this test
```

Comment:
This makes sense in the context of this PR, but will have no meaning some time in the future. Do we keep such comment?

Reply:
Well, this is a test and a delicate one, too. It was very hard to get a repro, so this comment is a word of caution for whoever is touching this. I could remove the sentence about my desktop, but I'd add a repro on how to break the code to trigger a failure.